src/platforms/linux/daemon/wireguardutilslinux.cpp

```diff
@@ -677,18 +689,66 @@ bool WireguardUtilsLinux::setupCgroupClass(const QString& path,
   return true;
 }
 
-QString WireguardUtilsLinux::getExcludeCgroup() const {
-  if (m_cgroups.isNull()) {
-    return QString();
+// static
+bool WireguardUtilsLinux::moveCgroupProcs(const QString& src,
+                                          const QString& dest) {
+  QFile srcProcs(src + "/cgroup.procs");
+  FILE* fp = fopen(qPrintable(dest + "/cgroup.procs"), "w");
+  if (!fp) {
+    return false;
+  }
+  auto guard = qScopeGuard([&] { fclose(fp); });
+
+  if (!srcProcs.open(QIODevice::ReadOnly | QIODevice::Text)) {
+    return false;
+  }
+  while (true) {
+    QString line = QString::fromLocal8Bit(srcProcs.readLine());
+    if (line.isEmpty()) {
+      break;
+    }
+    fputs(qPrintable(line), fp);
+    fflush(fp);
+  }
+  srcProcs.close();
+  return true;
+}
+
+void WireguardUtilsLinux::excludeCgroup(const QString& cgroup) {
+  logger.error() << "Excluding traffic from" << cgroup;
+  if (m_cgroupVersion == 1) {
+    // Add all PIDs from the unified cgroup to the net_cls exclusion cgroup.
+    moveCgroupProcs(m_cgroupUnified + cgroup,
+                    m_cgroupNetClass + VPN_EXCLUDE_CGROUP);
+  } else if (m_cgroupVersion == 2) {
+    QByteArray cgpath = cgroup.toLocal8Bit();
+    GoString goCgroup = {.p = cgpath.constData(),
+                         .n = (ptrdiff_t)cgpath.length()};
+    NetfilterMarkCgroupV2(goCgroup);
+  }
+}
+
+void WireguardUtilsLinux::resetCgroup(const QString& cgroup) {
+  logger.error() << "Permitting traffic from" << cgroup;
+  if (m_cgroupVersion == 1) {
+    // Add all PIDs from the unified cgroup to the net_cls default cgroup.
+    moveCgroupProcs(m_cgroupUnified + cgroup, m_cgroupNetClass);
+  } else if (m_cgroupVersion == 2) {
+    QByteArray cgpath = cgroup.toLocal8Bit();
+    GoString goCgroup = {.p = cgpath.constData(),
+                         .n = (ptrdiff_t)cgpath.length()};
+    NetfilterResetCgroupV2(goCgroup);
   }
```

Comment:
do you want to add an `} else { unexpected cgroup version` ?

Reply:
or a Q_ASSERT(false)

Reply:
There is also the possibility of `m_cgroupVersion == 0` indicating that the kernel doesn't support control groups at all.